### PR TITLE
`disk_encryption_set`: Support managed HSMs

### LINKED ISSUE #21470: disk_encryption_set doesn't support referencing a key in a managed HSM key vault in the key_vault_key_id
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.4.5

### AzureRM Provider Version

3.52.0

### Affected Resource(s)/Data Source(s)

azurerm_disk_encryption_set

### Terraform Configuration Files

```hcl
resource "azurerm_disk_encryption_set" "des" {
  name                = "des-test"
  resource_group_name = var.resource_group_name
  location            = var.location
  key_vault_key_id    = var.key_id
  encryption_type     = var.encryption_type

  identity {
    type         = "SystemAssigned"
  }
}
```

### Debug Output/Panic Output

```shell
Error: validating Key Vault Key "https://******.managedhsm.azure.net/keys/***/***" for Disk Encryption Set: retrieving the Resource ID the Key Vault at URL "https://******.managedhsm.azure.net/": expected a URI in the format `the-keyvault-name.vault.**` but got "******.managedhsm.azure.net"
```

### Expected Behaviour

The provider should allow for keys in managed HSMs to be used for disk encryption sets. Tested with the az cli and creating new des with managed HSM key URIs is supported.

### Actual Behaviour

Resource failed provider validation due to the logic in the base URL parsing:
https://github.com/hashicorp/terraform-provider-azurerm/blob/9a8299335488b18860282d391842e09c78d68ac5/internal/services/keyvault/client/helpers.go#L182

### Steps to Reproduce

1. Create a key in a managed HSM
2. Use that key in the azurerm_disk_encryption_set resource

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by Botje**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

This PR adds support for HSMs to the disk encryption set resource.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

I did not run any tests because deploying Managed HSMs exceeds the budget of my personal playground subscription.

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `disk_encryption_set` - support for managed HSMs [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #21470 

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Comment by Botje** [comment]
"service/disk" seems an appropriate label, too.

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @Botje - looks like we have a test failure:

```
------- Stdout: -------
=== RUN   TestAccDiskEncryptionSet_disablePurgeProtection
=== PAUSE TestAccDiskEncryptionSet_disablePurgeProtection
=== CONT  TestAccDiskEncryptionSet_disablePurgeProtection
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: validating Key Vault "acctestkv-jknzr" (Resource Group "acctestRG-240618213622132222") for Disk Encryption Set: Purge Protection must be enabled
        
          with azurerm_disk_encryption_set.test,
          on terraform_plugin_test.tf line 105, in resource "azurerm_disk_encryption_set" "test":
         105: resource "azurerm_disk_encryption_set" "test" {
        
--- FAIL: TestAccDiskEncryptionSet_disablePurgeProtection (328.83s)
FAIL
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @Botje - looks like we have a test failure:

```
------- Stdout: -------
=== RUN   TestAccDiskEncryptionSet_disablePurgeProtection
=== PAUSE TestAccDiskEncryptionSet_disablePurgeProtection
=== CONT  TestAccDiskEncryptionSet_disablePurgeProtection
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: validating Key Vault "acctestkv-jknzr" (Resource Group "acctestRG-240618213622132222") for Disk Encryption Set: Purge Protection must be enabled
        
          with azurerm_disk_encryption_set.test,
          on terraform_plugin_test.tf line 105, in resource "azurerm_disk_encryption_set" "test":
         105: resource "azurerm_disk_encryption_set" "test" {
        
--- FAIL: TestAccDiskEncryptionSet_disablePurgeProtection (328.83s)
FAIL
```

File: internal/services/compute/disk_encryption_set_data_source.go [hunk 1]
```diff
@@ -84,7 +86,16 @@ func dataSourceDiskEncryptionSetRead(d *pluginsdk.ResourceData, meta interface{}
 		d.Set("auto_key_rotation_enabled", props.RotationToLatestKeyVersionEnabled)
 
 		if props.ActiveKey != nil && props.ActiveKey.KeyUrl != "" {
-			d.Set("key_vault_key_url", props.ActiveKey.KeyUrl)
+			keyVaultURI := props.ActiveKey.KeyUrl
+			isHSMURI, err, _, _ := managedHsmHelpers.IsManagedHSMURI(env, keyVaultURI)
```
Comment (tombuildsstuff) [inline internal/services/compute/disk_encryption_set_data_source.go:90]: Heads up that this function is being removed in https://github.com/hashicorp/terraform-provider-azurerm/pull/26218 and @wuxu92 is adding a replacement for this in an upcoming PR - so this PR is dependent on https://github.com/hashicorp/terraform-provider-azurerm/pull/26218 / the PR from @wuxu92

**Comment by Botje** [comment]
Hi @katbyte,
> --- FAIL: TestAccDiskEncryptionSet_disablePurgeProtection (328.83s)

This was an asymmetry I noticed in the code: the creation path does not test for purge protection but the update path does. I assumed this was an oversight and implemented the same check for both create and update.
Looking at the git blame output for that specific test, I notice that that actually _removed_ the purge protection check on creation.
I think the proper course of action is then also to remove the purge protection check alltogether.

File: internal/services/compute/disk_encryption_set_data_source.go (hunk 1 above)
Comment (katbyte) [reply internal/services/compute/disk_encryption_set_data_source.go:90]: I am going to approve and merge this to expedite it in due to an internal ask. @wuxu92  can update it in #26218 with the pattern used to remove it

**Review by katbyte — APPROVED** [review]
Thanks @Botje that makes sense to me and all tests now pass. LGTM :basecamp:

**Pull request merged**